src/googleclouddebugger/firebase_client.py

```diff
@@ -0,0 +1,587 @@
+        bp_ref = firebase_admin.db.reference(
+            f'cdbg/breakpoints/{self._debuggee_id}/final/{bp_id}')
+        bp_ref.set(breakpoint_data)
+
+        native.LogInfo(f'Breakpoint {bp_id} update transmitted successfully')
+
+      # TODO: Add any firebase-related error handling.
+      except googleapiclient.errors.HttpError as err:
+        # Treat 400 error codes (except timeout) as application error that will
+        # not be retried. All other errors are assumed to be transient.
+        status = err.resp.status
+        is_transient = ((status >= 500) or (status == 408))
+        if is_transient:
+          if retry_count < self.max_transmit_attempts - 1:
+            native.LogInfo(f'Failed to send breakpoint {bp_id} update: '
+                           f'{traceback.format_exc()}')
+            retry_list.append((breakpoint_data, retry_count + 1))
+          else:
+            native.LogWarning(
+                f'Breakpoint {bp_id} retry count exceeded maximum')
+        else:
+          # This is very common if multiple instances are sending final update
+          # simultaneously.
+          native.LogInfo(f'{err}, breakpoint: {bp_id}')
+      except socket.error as err:
+        if retry_count < self.max_transmit_attempts - 1:
+          native.LogInfo(f'Socket error {err.errno} while sending breakpoint '
+                         f'{bp_id} update: {traceback.format_exc()}')
+          retry_list.append((breakpoint_data, retry_count + 1))
+        else:
+          native.LogWarning(f'Breakpoint {bp_id} retry count exceeded maximum')
+          # Socket errors shouldn't persist like this; reconnect.
+          #reconnect = True
+      except BaseException:
+        native.LogWarning(f'Fatal error sending breakpoint {bp_id} update: '
+                          f'{traceback.format_exc()}')
+
+    self._transmission_queue.extend(retry_list)
+
+    if not self._transmission_queue:
+      self.update_backoff.Succeeded()
+      # Nothing to send, wait until next breakpoint update.
+      return None
+    else:
+      return self.update_backoff.Failed()
+
+  def _GetDebuggee(self):
+    """Builds the debuggee structure."""
+    major_version = 'v' + version.__version__.split('.', maxsplit=1)[0]
+    python_version = ''.join(platform.python_version().split('.')[:2])
+    agent_version = f'google.com/python{python_version}-gcp/{major_version}'
+
+    debuggee = {
+        'description': self._GetDebuggeeDescription(),
+        'labels': self._debuggee_labels,
+        'agentVersion': agent_version,
+        'canaryMode': self._canary_mode,
+    }
+
+    source_context = self._ReadAppJsonFile('source-context.json')
+    if source_context:
+      debuggee['sourceContexts'] = [source_context]
+
+    debuggee['uniquifier'] = self._ComputeUniquifier(debuggee)
+
+    debuggee['id'] = self._ComputeDebuggeeId(debuggee)
+
+    return debuggee
+
+  def _ComputeDebuggeeId(self, debuggee):
```

Comment:
There should probably be a comment about uniqueness, similar to the comment for uniquifier. Also, should any fields/data be stripped from the computation? Even with labels and region, I guess we want to include that too right?

Reply:
Added a comment.

Good question about whether the computation should be limited.  I've left it with the simple implementation right now, but there may be a need (or desire) to alter that in the future depending on how labels are being used.  I believe that the current label set are stable and do not contain anything that would inadvertently split debuggees.

Reply:
The main thing that would split them is the region I think if it's Cloud Function, but in that case I believe we do want them to have different debuggee IDs?

Reply:
With the general push towards regionalization, yes, I'd have them be different debuggee IDs by default.  It could be useful to merge them together in the future, but I'd consider that to be a feature request.